src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs

```diff
@@ -192,6 +221,10 @@ public T GetValue(CancellationToken cancellationToken)
                     return _cachedResult.Result;
                 }
 
+                Contract.ThrowIfTrue(
+                    forceSynchronousComputeOnCallingThread && _synchronousComputeFunction is null,
+                    "Should never ask to compute the value synchronously when there is no synchronous function to do the work.");
```

Comment:
note: this check needs to happen here as it must be under the lock (since we mutate our functions by setting them to null), and must be after the inner lock check to see if we've transitioned to the 'complete' state.